Wired up storage of EmailSpamComplaintEvents

allouis opened the pull request:
refs https://github.com/TryGhost/Team/issues/2337

This also removes the code which unsubscribes members from newsletters because that is not the spec of the feature.

allouis committed:
Wired up storage of EmailSpamComplaintEvents

refs https://github.com/TryGhost/Team/issues/2337

This also removes the code which unsubscribes members from newsletters because
that is not the spec of the feature.

allouis committed:
Added support for EmailSpamComplaintEvent to activity feed

We need to add relations and filterRelations to the model so that we can
correctly filter events by post_id and member_id

SimonBackx commented on ghost/email-service/lib/email-event-storage.js:
```diff
@@ -61,7 +66,11 @@ class EmailEventStorage {
 
     async handleComplained(event) {
-        return this.unsubscribeFromNewsletters(event);
+        await this.#EmailSpamComplaintEvent.add({
```
Awaiting discussion about duplicate events, this might need protection to prevent adding the same event multiple times in the db

allouis replied:
Going to add protection for now - good shout!

SimonBackx commented on ghost/email-service/lib/email-event-storage.js:
```diff
@@ -61,7 +66,11 @@ class EmailEventStorage {
     }
 
     async handleComplained(event) {
-        return this.unsubscribeFromNewsletters(event);
```
Do we want a feature flag check here?

allouis replied:
Discussed in slack: we're good to store the new data without a flag check, as it'll give us a backlog of events if release is delayed.

allouis committed:
Added unique constraint to email_spam_complaint_events table

We can fetch the same event multiple times from Mailgun so we need to
be able to protect against inserting duplicate events in the
database. This will allow us to catch duplicate errors on insert when
handling complaint events.

allouis committed:
Handled receiving duplicate SpamComplaintEvents

We need to check for both ER_DUP_ENTRY and SQLITE_CONSTRAINT because
different DB engines give us back different error codes.

allouis commented:
@daniellockyer I've updated the schema and a migration to add a unique constraint to the email_spam_complaint_events table, so that we can ensure that duplicate events do not get inserted.

daniellockyer reviewed (approved):
@allouis The change is fine with me but this won't apply retroactively to developers who have already run the migration, so please keep this in mind when it comes to bugs found during development
